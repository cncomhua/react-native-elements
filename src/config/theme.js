// For testing purposes only
import colors from './colors';

export default {
  colors,
};
