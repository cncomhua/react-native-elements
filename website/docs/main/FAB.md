---


id: fab


title: FAB


slug: /fab

---



import Usage from './usage/FAB/FAB.md'



A floating action button (FAB) performs the primary, or most common, action on a screen.

It appears in front of all screen content, typically as a circular shape with an icon in its center.

Also receives all [Button](https://reactnativeelements.com/docs/button#props) props.



## Usage


<Usage />

---


## Props

### FAB




## Reference

### FAB

