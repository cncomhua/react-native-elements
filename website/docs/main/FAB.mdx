---
id: fab

title: FAB

slug: /fab
---

import Usage from './usage/FAB/FAB.mdx';

import Props from './props/FAB.mdx';

import { FAB } from 'react-native-elements';

import Tabs from '@theme/Tabs';

import TabItem from '@theme/TabItem';

A floating action button (FAB) performs the primary, or most common, action on a screen.

It appears in front of all screen content, typically as a circular shape with an icon in its center.

Also receives all [Button](https://reactnativeelements.com/docs/button#props) props.

## Usage

<Usage />

## Props

<Props />
