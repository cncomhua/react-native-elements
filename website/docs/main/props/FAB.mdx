### FAB

- [color](#color)
- [placement](#placement)
- [size](#size)
- [style](#style)
- [upperCase](#uppercase)
- [visible](#visible)

## Reference

### FAB

#### color

Change the color of the FAB.

| Type   | Default |
| ------ | ------- |
| string | None    |

---

#### placement

FAB placement at bottom, (optional) use [`style`](#style) in case of custom placement.

| Type              | Default |
| ----------------- | ------- |
| "left" or "right" | None    |

---

#### size

Change Size of FAB.

| Type               | Default |
| ------------------ | ------- |
| "small" or "large" | large   |

---

#### style

//FIXME: not in doc but available in examples

Style for FAB

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### upperCase

Transform Extended Label text to uppercase.

| Type    | Default |
| ------- | ------- |
| boolean | None    |

---

#### visible

Decide the visibility of the FAB.

| Type    | Default |
| ------- | ------- |
| boolean | true    |

---
