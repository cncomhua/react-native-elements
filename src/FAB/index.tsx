import { withTheme } from '../config';
import { FAB, FABProps } from './FAB';

export { FAB };
export type { FABProps };
/**
 * RNE FAB
 *
 * **Preview**
 *
 * ![URL](URL)
 *
 * [**Documentation**](https://reactnativeelements.com/docs/fab)
 */
export default withTheme(FAB, 'FAB');
