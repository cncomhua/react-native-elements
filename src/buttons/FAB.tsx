import React from 'react';
import { StyleSheet, Animated, StyleProp, ViewStyle } from 'react-native';
import Button, { ButtonProps } from './Button';
import { IconProps } from '../icons/Icon';
import { FullTheme, withTheme } from '../config';

export type FABProps = ButtonProps & {
  iconName?: string;
  iconColor?: string;
  iconProps?: IconProps;
  color?: string;
  label?: string;
  labelColor?: string;
  size?: 'large' | 'small';
  placement?: 'left' | 'right';
  visible?: boolean;
  disabled?: boolean;
  upperCase?: boolean;
  style?: StyleProp<ViewStyle>;
  theme?: FullTheme;
};

const FAB: React.FunctionComponent<FABProps> = ({
  iconName,
  iconColor,
  iconProps,
  color,
  label,
  labelColor,
  size = 'large',
  visible = true,
  disabled,
  upperCase,
  theme,
  style,
  titleStyle,
  buttonStyle,
  containerStyle,
  iconContainerStyle,
  placement,
  ...props
}) => {
  const { current: animation } = React.useRef(
    new Animated.Value(Number(visible))
  );

  React.useEffect(() => {
    Animated.timing(animation, {
      toValue: Number(visible),
      duration: 200,
      useNativeDriver: true,
    }).start();
  }, [animation, visible]);

  return (
    <Animated.View
      style={[
        {
          opacity: animation,
          transform: [{ scale: animation }],
        },
        styles.content,
        placement && {
          position: 'absolute',
          margin: 16,
          [placement]: 0,
          bottom: 0,
        },
        style,
      ]}
    >
      <Button
        icon={{ name: iconName, color: iconColor || theme.colors.white }}
        {...props}
        buttonStyle={StyleSheet.flatten([
          label
            ? styles.extendedLabel
            : size === 'small'
            ? styles.smallFAB
            : styles.largeFAB,
          {
            backgroundColor: color || theme.colors.primary,
          },
          buttonStyle,
        ])}
        iconContainerStyle={[
          label ? {} : size === 'small' ? styles.smallFAB : styles.largeFAB,
          iconContainerStyle,
        ]}
        containerStyle={StyleSheet.flatten([
          styles.container,
          disabled && { elevation: 0 },
          containerStyle,
        ])}
        disabled={disabled}
        title={label}
        titleStyle={[
          styles.label,
          { color: labelColor || theme.colors.white },
          upperCase && styles.upperCaseLabel,
          titleStyle,
        ]}
      />
    </Animated.View>
  );
};
const styles = StyleSheet.create({
  container: {
    elevation: 4,
    borderRadius: 28,
  },
  largeFAB: {
    height: 56,
    width: 56,
    padding: 16,
  },
  smallFAB: {
    height: 40,
    width: 40,
    padding: 8,
  },
  content: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 28,
  },
  label: {
    marginHorizontal: 8,
  },
  upperCaseLabel: {
    textTransform: 'uppercase',
  },
  extendedLabel: {
    height: 48,
    paddingHorizontal: 16,
  },
});
export { FAB };

export default withTheme(FAB, 'FloatingActionButton');
