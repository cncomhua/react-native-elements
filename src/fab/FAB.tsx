import React from 'react';
import {
  TouchableOpacity,
  View,
  StyleSheet,
  Animated,
  Text,
  StyleProp,
  ViewStyle,
  ViewProps,
} from 'react-native';
import { Icon, IconProps } from '../icons/Icon';
import { FullTheme } from '../config';

type FABProps = ViewProps & {
  icon?: string;
  iconColor?: string;
  iconProps?: IconProps;
  IconComponent?: JSX.Element;
  color?: string;
  label?: string;
  labelColor?: string;
  size?: 'large' | 'small';
  visible?: boolean;
  disabled?: boolean;
  upperCase?: boolean;
  style?: StyleProp<ViewStyle>;
  onPress?: () => void;
  onLongPress?: () => void;
  theme?: FullTheme;
};

const FAB: React.FunctionComponent<FABProps> = ({
  icon,
  iconColor,
  iconProps,
  IconComponent,
  color,
  label,
  labelColor,
  size = 'large',
  visible = true,
  disabled,
  upperCase,
  theme,
  style,
  onPress,
  onLongPress,
  ...props
}) => {
  const { current: animation } = React.useRef(
    new Animated.Value(Number(visible))
  );

  React.useEffect(() => {
    Animated.timing(animation, {
      toValue: Number(visible),
      duration: 200,
      useNativeDriver: true,
    }).start();
  }, [animation, visible]);

  return (
    <Animated.View
      {...props}
      style={[
        {
          opacity: animation,
          transform: [{ scale: animation }],
        },
        styles.container,
        style,
      ]}
    >
      <TouchableOpacity
        disabled={disabled}
        style={styles.touchable}
        onPress={disabled ? () => {} : onPress}
        onLongPress={disabled ? () => {} : onLongPress}
      >
        <View
          style={StyleSheet.flatten([
            styles.content,
            {
              backgroundColor: color || theme.colors.primary,
            },
            disabled && {
              backgroundColor: theme.colors.grey4,
            },
            label
              ? styles.extendedLabel
              : size === 'small'
              ? styles.smallFAB
              : styles.largeFAB,
          ])}
        >
          {icon ? (
            <Icon
              {...iconProps}
              name={icon}
              color={iconColor || theme.colors.secondary}
            />
          ) : (
            Boolean(IconComponent) && IconComponent
          )}
          {Boolean(label) && (
            <Text
              selectable={false}
              style={[
                styles.label,
                { color: labelColor || theme.colors.white },
                upperCase && styles.upperCaseLabel,
              ]}
            >
              {label}
            </Text>
          )}
        </View>
      </TouchableOpacity>
    </Animated.View>
  );
};
const styles = StyleSheet.create({
  container: {
    elevation: 10,
  },
  touchable: {
    borderRadius: 28,
  },
  largeFAB: {
    height: 56,
    width: 56,
    padding: 16,
  },
  smallFAB: {
    height: 40,
    width: 40,
    padding: 8,
  },
  content: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 28,
  },
  label: {
    marginHorizontal: 8,
  },
  upperCaseLabel: {
    textTransform: 'uppercase',
  },
  extendedLabel: {
    height: 48,
    paddingHorizontal: 16,
  },
});

export default FAB;
