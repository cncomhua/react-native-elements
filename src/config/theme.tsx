import { ViewProps } from 'react-native';
import { SwipeRatingProps } from '../Rating';
import { TapRatingProps } from '../AirbnbRating';
import { AvatarProps, AccessoryProps } from '../Avatar';
import { BadgeProps } from '../Badge';
import { BottomSheetProps } from '../BottomSheet';
import { ButtonProps } from '../Button';
import { ButtonGroupProps } from '../ButtonGroup';
import { CardProps } from '../card/Card';
import { CheckBoxProps } from '../Checkbox';
import { DividerProps } from '../Divider';
import { HeaderProps } from '../Header';
import { IconProps } from '../Icon';
import { ImageProps } from '../Image';
import { InputProps } from '../Input';
import { ListItemProps } from '../ListItem';
import { ListItemAccordionProps } from '../ListItem';
import { OverlayProps } from '../Overlay';
import { PricingCardProps } from '../PricingCard';
import { SearchBarProps } from '../searchbar/SearchBar';
import { SliderProps } from '../Slider';
import { SocialIconProps } from '../SocialIcon';
import { TextProps } from '../Text';
import { TileProps } from '../Tile';
import { TooltipProps } from '../Tooltip';
import { SwitchProps } from '../Switch';
import { TabItemProps, TabProps } from '../tab/Tab';
import { TabViewProps } from '../tab/TabView';
import { FABProps } from '../FAB';
import { SpeedDialProps, SpeedDialActionProps } from '../SpeedDial';
import { LinearProgressProps } from '../LinearProgress';
import { ChipProps } from '../Chip';
import {
  DialogActionsProps,
  DialogLoadingProps,
  DialogProps,
  DialogTitleProps,
} from '../Dialog';

import colors, { Colors } from './colors';

export default {
  colors,
};

type RecursivePartial<T> = { [P in keyof T]?: RecursivePartial<T[P]> };

export interface FullTheme {
  Avatar: Partial<AvatarProps>;
  AvatarAccessory: Partial<AccessoryProps>;
  Badge: Partial<BadgeProps>;
  BottomSheet: Partial<BottomSheetProps>;
  Button: Partial<ButtonProps>;
  ButtonGroup: Partial<ButtonGroupProps>;
  Chip: Partial<ChipProps>;
  Card: Partial<CardProps>;
  CardDivider: Partial<DividerProps>;
  CardFeaturedSubtitle: Partial<TextProps>;
  CardFeaturedTitle: Partial<TextProps>;
  CardImage: Partial<ImageProps>;
  CardTitle: Partial<TextProps>;
  CheckBox: Partial<CheckBoxProps>;
  Divider: Partial<DividerProps>;
  Dialog: Partial<DialogProps>;
  DialogActions: Partial<DialogActionsProps>;
  DialogButton: Partial<ButtonProps>;
  DialogLoading: Partial<DialogLoadingProps>;
  DialogTitle: Partial<DialogTitleProps>;
  Header: Partial<HeaderProps>;
  Icon: Partial<IconProps>;
  Image: Partial<ImageProps>;
  Input: Partial<InputProps>;
  ListItem: Partial<ListItemProps>;
  ListItemAccordion: Partial<ListItemAccordionProps>;
  ListItemButtonGroup: Partial<ButtonGroupProps>;
  ListItemCheckBox: Partial<CheckBoxProps>;
  ListItemContent: Partial<ViewProps>;
  ListItemChevron: Partial<IconProps>;
  ListItemInput: Partial<InputProps>;
  ListItemSubtitle: Partial<TextProps>;
  ListItemTitle: Partial<TextProps>;
  Overlay: Partial<OverlayProps>;
  PricingCard: Partial<PricingCardProps>;
  Rating: Partial<SwipeRatingProps>;
  AirbnbRating: Partial<TapRatingProps>;
  SearchBar: Partial<SearchBarProps>;
  Slider: Partial<SliderProps>;
  SocialIcon: Partial<SocialIconProps>;
  Text: Partial<TextProps>;
  Tile: Partial<TileProps>;
  Switch: Partial<SwitchProps>;
  Tooltip: Partial<TooltipProps>;
  colors: RecursivePartial<Colors>;
  Tab: Partial<TabProps>;
  TabItem: Partial<TabItemProps>;
  TabView: Partial<TabViewProps>;
  TabViewItem: Partial<ViewProps>;
  FAB: Partial<FABProps>;
  SpeedDial: Partial<SpeedDialProps>;
  SpeedDialAction: Partial<SpeedDialActionProps>;
  LinearProgress: Partial<LinearProgressProps>;
}

export type Theme<T = {}> = Partial<FullTheme> & T;

export type UpdateTheme = (updates: RecursivePartial<FullTheme>) => void;

export type ReplaceTheme = (updates: RecursivePartial<FullTheme>) => void;
