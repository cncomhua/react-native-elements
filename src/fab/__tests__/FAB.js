import React from 'react';
import FAB from '../FAB';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import theme from '../../config/theme';

describe('LinearProgress Component', () => {
  it('should render without issues', () => {
    const app = shallow(<FAB theme={theme} />);
    expect(app.length).toBe(1);
    expect(toJson(app)).toMatchSnapshot();
  });
});
