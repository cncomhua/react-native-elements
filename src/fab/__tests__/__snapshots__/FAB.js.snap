// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinearProgress Component should render without issues 1`] = `
<ForwardRef(AnimatedComponentWrapper)
  style={
    Array [
      Object {
        "opacity": 1,
        "transform": Array [
          Object {
            "scale": 1,
          },
        ],
      },
      Object {
        "elevation": 10,
      },
      undefined,
    ]
  }
>
  <ForwardRef
    style={
      Object {
        "borderRadius": 28,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#2089dc",
          "borderRadius": 28,
          "flexDirection": "row",
          "height": 56,
          "justifyContent": "center",
          "padding": 16,
          "width": 56,
        }
      }
    />
  </ForwardRef>
</ForwardRef(AnimatedComponentWrapper)>
`;
